indico_toolkit/association/association.py

```diff
@@ -0,0 +1,59 @@
+from typing import Union, List, Dict
+from collections import defaultdict
+from indico_toolkit.types import Predictions
+
+
+class Association:
+    """
+    Base class for matching tokens to extraction predictions. 
+    """
+
+    def __init__(self, predictions: Union[List[dict], Predictions]):
+        self._predictions = self.validate_prediction_formatting(predictions)
+        self._mapped_positions = []
+        self._manually_added_preds = []
+        self._errored_predictions = []
+
+    @staticmethod
+    def match_pred_to_token(self):
+        raise NotImplementedError
+
+    @staticmethod
+    def sort_predictions_by_start_index(predictions: List[dict]) -> List[dict]:
+        return sorted(predictions, key=lambda x: x["start"])
+
+    @property
+    def mapped_positions_by_page(self) -> Dict[int, List[dict]]:
+        """
+        Return mapped positions by page on which they first appear
+        """
+        page_map = defaultdict(list)
+        for position in self._mapped_positions:
+            page_map[position["page_num"]].append(position)
+        return page_map
+    
+    def validate_prediction_formatting(
+        self, predictions: Union[List[dict], Predictions]
+    ) -> List[dict]:
+        if isinstance(predictions, Predictions):
+            predictions = predictions.to_list()
+        return predictions
+
+    def _is_manually_added(self, prediction: dict):
```

Comment:
this function (given that we're passing a mutable by reference) has a side-effect on the prediction.  is there a way to make that explicit...perhaps the function name?

Reply:
error key is being added outside of this method now